covalent_indexer.go

```diff
@@ -31,91 +39,196 @@ func NewCovalentDataIndexer(processor DataHandler, server *http.Server) (*covale
+			wsr = ci.getWSR()
+
+			if wss != nil && wsr != nil {
+				dataSent := ci.sendDataWithAcknowledge(data, ackData, wss, wsr)
+				if dataSent {
+					return
+				}
+			}
+		}
+	}
+}
 
-	if wss != nil {
-		//TODO: Handle error in next PR
-		_ = wss.SendMessage(binaryData)
+func (ci *covalentIndexer) sendDataWithAcknowledge(
+	data []byte,
+	ackData []byte,
+	wss process.WSConn,
+	wsr process.WSConn,
+) bool {
+	errSend := wss.WriteMessage(websocket.BinaryMessage, data)
+	if errSend != nil {
+		log.Warn("could not send block data to covalent, waiting for new connection", "error", errSend)
+		ci.waitForWSSConnection()
 	}
+
+	msgType, receivedData, errReadData := wsr.ReadMessage()
+	if errReadData != nil {
+		log.Warn("could not receive acknowledge data from covalent, waiting for new connection", "error", errReadData)
+		ci.waitForWSRConnection()
+	}
+
+	if errSend == nil && errReadData == nil && msgType == websocket.BinaryMessage {
+		if bytes.Compare(receivedData, ackData) == 0 {
+			return true
+		}
+	}
+
+	return false
 }
 
 // SaveBlock saves the block info and converts it in order to be sent to covalent
-func (c *covalentIndexer) SaveBlock(args *indexer.ArgsSaveBlockData) {
-	blockResult, err := c.processor.ProcessData(args)
+func (ci *covalentIndexer) SaveBlock(args *indexer.ArgsSaveBlockData) {
+	blockResult, err := ci.processor.ProcessData(args)
 	if err != nil {
 		log.Error("SaveBlock failed. Could not process block",
 			"error", err, "headerHash", hex.EncodeToString(args.HeaderHash))
-		panic("could not process block, please check log")
+		panic("could not process block, check log")
 	}
 
-	c.sendBlockResultToCovalent(blockResult)
+	dataToSend, err := utility.Encode(blockResult)
+	if err != nil {
+		log.Error("could not encode block result to binary data", "error", err)
+		panic("could not encode block result, check log")
+	}
+
+	ci.sendWithRetrial(dataToSend, blockResult.Block.Hash)
 }
 
 // RevertIndexedBlock DUMMY
-func (c covalentIndexer) RevertIndexedBlock(header data.HeaderHandler, body data.BodyHandler) {}
+func (ci *covalentIndexer) RevertIndexedBlock(data.HeaderHandler, data.BodyHandler) {}
 
 // SaveRoundsInfo DUMMY
-func (c covalentIndexer) SaveRoundsInfo(roundsInfos []*indexer.RoundInfo) {}
+func (ci *covalentIndexer) SaveRoundsInfo([]*indexer.RoundInfo) {}
 
 // SaveValidatorsPubKeys DUMMY
-func (c covalentIndexer) SaveValidatorsPubKeys(validatorsPubKeys map[uint32][][]byte, epoch uint32) {}
+func (ci *covalentIndexer) SaveValidatorsPubKeys(map[uint32][][]byte, uint32) {
+}
 
 // SaveValidatorsRating DUMMY
-func (c covalentIndexer) SaveValidatorsRating(indexID string, infoRating []*indexer.ValidatorRatingInfo) {
+func (ci *covalentIndexer) SaveValidatorsRating(string, []*indexer.ValidatorRatingInfo) {
 }
 
 // SaveAccounts DUMMY
-func (c covalentIndexer) SaveAccounts(blockTimestamp uint64, acc []data.UserAccountHandler) {}
+func (ci *covalentIndexer) SaveAccounts(uint64, []data.UserAccountHandler) {}
 
 // Close DUMMY
```

Comment:
update comment

Reply:
Done in previous PR, merged the changes in this as well